exporter/otlphttpexporter/otlp_test.go

```diff
@@ -677,3 +641,231 @@ func TestUserAgent(t *testing.T) {
+			// `-1` indicates a missing Content-Length header in the Go http standard library
+			ContentLength: -1,
+			Body:          io.NopCloser(bytes.NewReader(data)),
+		}
+		err = handlePartialSuccessResponse(resp, tracesPartialSuccessHandler)
+		assert.True(t, consumererror.IsPermanent(err))
+	})
+
+	t.Run("Response is missing a Content-Length header and a body", func(t *testing.T) {
+		resp := &http.Response{
+			// `-1` indicates a missing Content-Length header in the Go http standard library
+			ContentLength: -1,
+			Body:          io.NopCloser(bytes.NewReader([]byte{})),
+		}
+		err := handlePartialSuccessResponse(resp, tracesPartialSuccessHandler)
+		assert.Nil(t, err)
+	})
+
+	t.Run("Reading the response body returns an error other than ErrUnexpectedEOF", func(t *testing.T) {
+		resp := &http.Response{
+			// `-1` indicates a missing Content-Length header in the Go http standard library
+			ContentLength: -1,
+			Body:          io.NopCloser(badReader{}),
+		}
+		err := handlePartialSuccessResponse(resp, tracesPartialSuccessHandler)
+		assert.Error(t, err)
+	})
+
+	t.Run("Response has short Content-Length header", func(t *testing.T) {
+		response := ptraceotlp.NewExportResponse()
+		partial := response.PartialSuccess()
+		partial.SetErrorMessage("hello")
+		partial.SetRejectedSpans(1)
+		data, err := response.MarshalProto()
+		require.NoError(t, err)
+		resp := &http.Response{
+			ContentLength: 3,
+			Body:          io.NopCloser(bytes.NewReader(data)),
+		}
+		err = handlePartialSuccessResponse(resp, tracesPartialSuccessHandler)
+		assert.Error(t, err)
+	})
+
+	t.Run("Response has long Content-Length header", func(t *testing.T) {
+		response := ptraceotlp.NewExportResponse()
+		partial := response.PartialSuccess()
+		partial.SetErrorMessage("hello")
+		partial.SetRejectedSpans(1)
+		data, err := response.MarshalProto()
+		require.NoError(t, err)
+		resp := &http.Response{
+			ContentLength: 4096,
+			Body:          io.NopCloser(bytes.NewReader(data)),
+		}
+		err = handlePartialSuccessResponse(resp, tracesPartialSuccessHandler)
+		assert.Error(t, err)
+	})
+
+	invalidBodyCases := []struct {
+		telemetryType string
+		handler       partialSuccessHandler
+	}{
+		{
+			telemetryType: "traces",
+			handler:       tracesPartialSuccessHandler,
+		},
+		{
+			telemetryType: "metrics",
+			handler:       metricsPartialSuccessHandler,
+		},
+		{
+			telemetryType: "logs",
+			handler:       logsPartialSuccessHandler,
+		},
+	}
+	for _, tt := range invalidBodyCases {
+		t.Run("Invalid response body: "+tt.telemetryType, func(t *testing.T) {
+			str := "invalid proto"
+			body := bytes.NewBufferString(str)
+			resp := &http.Response{
+				ContentLength: int64(len(str)),
+				Body:          io.NopCloser(body),
+			}
+			err := handlePartialSuccessResponse(resp, tt.handler)
+			assert.Error(t, err)
+		})
+	}
+
+}
+
+func createBackend(addr string, endpoint string, handler func(writer http.ResponseWriter, request *http.Request)) (*http.Server, error) {
```

Comment:
Thanks for the tip. I've adapted the tests to use `httptest.Server`, but it felt like a wash overall since the server URL is used when checking error conditions. Let me know what you think.